docs/content/integrations/dbt/using-dbt-with-dagster/upstream-assets.mdx

```diff
@@ -15,127 +15,140 @@ description: Dagster can orchestrate dbt alongside other technologies.
+    # Log some metadata about the table we just wrote. It will show up in the UI.
+    context.add_output_metadata({"num_rows": data.shape[0]})
 
-   @asset(key_prefix=["jaffle_shop"])
-   def orders_raw() -> pd.DataFrame:
-       data = pd.read_csv("https://docs.dagster.io/assets/orders.csv")
-       return data
-   ```
 
-Let's take a closer look at the argument we've provided:
+dbt = DbtCliResource(project_dir=os.fspath(dbt_project_dir))
+```
 
-- `key_prefix` - When the assets are materialized, Dagster will store them in DuckDB in the schema defined by the last value in `key_prefix`. In this case, that's `jaffle_shop`. The tables will have the same names as the assets that produced them, which are `customers_raw` and `orders_raw`.
+Let's review the changes we made:
 
-  Because these tables will become the source data for the `stg_customers.sql` and `stg_orders.sql` models in the dbt project, the names of the assets must match the table names specified in `/tutorial_template/jaffle_shop/models/sources.yml`, which you configured in [part one of this tutorial](/integrations/dbt/using-dbt-with-dagster/set-up-dbt-project#step-3-configure-dbt-model-data-sources).
+1. At the top, we added imports for `pandas` and `duckdb`, which we use to for fetching data into a `DataFrame` and writing it to DuckDB.
 
-At this point, the `/tutorial_template/tutorial_dbt_dagster/assets/__init__.py` file should look like this:
+2. We added a `duckdb_database_path` variable, which holds the location of our DuckDB database. Remember that DuckDB databases are just regular files on the local filesystem. The path is the same path that we used when we configured our profiles.yml file. This variable is used in the implementations of the `raw_customers` asset.
 
-```python
-import pandas as pd
-from dagster_dbt import load_assets_from_dbt_project
+3. We added a definition for the `raw_customers` table. The `raw_customers` function with the <PyObject object="asset" decorator /> decorator above it is the function that defines the `raw_customers` table. The implementation inside the function fetches data from the internet and writes it to a table in our DuckDB database. Similar to how running a dbt model executes a select statement, materializing this asset will execute this Python code.
 
-from dagster import asset, file_relative_path
+Finally, let's update the `assets` argument of our `Definitions` object, at the bottom of the file, to include the new asset we just defined:
 
+```python file=integrations/dbt/tutorial/upstream_assets/definitions.py startafter=start_defs endbefore=end_defs
+defs = Definitions(
+    assets=[raw_customers, jaffle_shop_dbt_assets],
+    schedules=schedules,
+    resources={
+        "dbt": dbt,
+    },
+)
+```
 
-@asset(key_prefix=["jaffle_shop"])
-def customers_raw() -> pd.DataFrame:
-   data = pd.read_csv("https://docs.dagster.io/assets/customers.csv")
-   return data
+---
 
+## Step 2: In the dbt project, replace a seed with a source
```

Comment:
The code blocks are rendering funny in this whole section, likely because they're not indented under the line items correctly. Could you address this? LMK if you need help, since Markdown can be finicky about this